File: frontend/src/lib/studentPortal.ts
// Student portal attendance import, via a REAL in-app WebView login.
//
// Why this exists: academia stopped publishing attendance (Aug 2026) while the
// student portal (sp.srmist.edu.in) kept it. The portal refuses scripted
// logins behind an anti-bot fingerprint gate that Skipp will not forge, so the
// student signs in for real inside a native WebView: real page, real captcha,
// real human. Because that WebView is same-origin with the portal once signed
// in, injected JavaScript can fetch the report pages directly and their session
// cookies (HttpOnly JSESSIONID included) ride along on their own. The HTML
// comes back to the app, which sends it to the backend to be parsed.
//
// NATIVE ONLY. On plain web (the PWA) the Capacitor bridge is absent, so this
// no-ops and the UI must gate on `canImportStudentPortal()`. There is no web
// fallback by design: a browser cannot read a cross-origin authenticated
// session, which is the whole reason this needs a native shell.
//
// NOT YET VERIFIED ON DEVICE. The backend parser and the /sp/parse route are
// tested against real data; this WebView bridge is written against the plugin's
// documented API and must be run in an actual iOS/Android build to confirm.

import { Capacitor } from "@capacitor/core";
import { InAppBrowser } from "@capgo/inappbrowser";

import { parseStudentPortal } from "@/lib/api";
import type { StudentPortalSnapshot } from "@/types";

const PORTAL_ORIGIN = "https://sp.srmist.edu.in";
const CONTEXT = "/srmiststudentportal";
const LOGIN_URL = `${PORTAL_ORIGIN}${CONTEXT}/students/template/HRDSystem.jsp`;
const REPORT_ATTENDANCE = `${CONTEXT}/students/report/studentAttendanceDetails.jsp`;
const REPORT_MARKS = `${CONTEXT}/students/report/studentInternalMarkDetails.jsp`;

// A tag on the message so a stray postMessage from the portal's own JS can
// never be mistaken for our capture.
const MSG = "skipp-sp-capture";

// The Android System WebView identifies itself with WebView markers in its
// User-Agent ("; wv" / "Version/4.0"), which the portal's anti-bot layer flags:
// the login page renders but the captcha image is refused, so sign-in is
// impossible. Overriding the UA to a plain Chrome-on-Android string makes the
// WebView present as an ordinary mobile browser, which is what it is (a real
// person is about to type into it). iOS WKWebView already passes as clean
// Safari, but we set the same UA on both so the two behave identically.
const PORTAL_UA =
  "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 " +
  "(KHTML, like Gecko) Chrome/126.0.0.0 Mobile Safari/537.36";

/** True only in the native shell, where a real WebView login is possible. */
export function canImportStudentPortal(): boolean {
  return Capacitor.isNativePlatform();
}

/** The student closed the login before signing in. Expected, not an error. */
export class ImportCancelled extends Error {
  readonly kind = "cancelled";
}

// Injected into the portal's page. It fetches the report same-origin (so
// cookies, including HttpOnly, are sent automatically) and reports back ONLY
// once the response is a real attendance table.
//
// Gating on a POSITIVE signal (the attendance table's own column headers) is
// deliberate and load-bearing: a signed-out fetch does not return the login
// form, it returns a tiny "Please wait login screen is loading" LOADER page
// that carries none of the usual login markers. An earlier "not signed out"
// gate therefore posted that empty loader the instant the WebView opened and
// closed before the student could type anything. Requiring the table means the
// probe stays quiet through every login redirect and fires exactly once, when
// real data finally appears. Runs on each navigation until then.
function probeScript(): string {
  return `
    (async () => {
      if (window.__skippCapturing) return;
      // The attendance report's own header cells. Present only on the real,
      // signed-in report, never on the login or loader pages.
      const hasAttendance = (h) => /Att\\.?\\s*hours|Total\\s*Percentage|Absent\\s*hours/i.test(h);
      try {
        const a = await fetch(${JSON.stringify(REPORT_ATTENDANCE)}, {
          method: "POST", credentials: "include",
        });
        const at = await a.text();
        if (!hasAttendance(at)) return; // not signed in yet, keep waiting
        window.__skippCapturing = true;
        let mt = null;
        try {
          const m = await fetch(${JSON.stringify(REPORT_MARKS)}, {
            method: "POST", credentials: "include",
          });
          mt = await m.text();
        } catch (e) {}
        window.mobileApp.postMessage({
          detail: { tag: ${JSON.stringify(MSG)}, attendanceHtml: at, marksHtml: mt },
        });
      } catch (e) {}
    })();
  `;
}

/**
 * Open the portal login, wait for a real sign-in, capture attendance (and
 * marks if present), and return them parsed.
 *
 * Rejects with `ImportCancelled` if the student closes the login first.
 */
export async function importStudentPortal(): Promise<StudentPortalSnapshot> {
  if (!canImportStudentPortal()) {
    throw new Error("Student portal import is only available in the Skipp app.");
  }

  const probe = probeScript();
  let done = false;

  const listeners: Array<{ remove: () => Promise<void> }> = [];
  const cleanup = async () => {
    for (const l of listeners) await l.remove().catch(() => {});
    await InAppBrowser.close().catch(() => {});
  };

  return new Promise<StudentPortalSnapshot>((resolve, reject) => {
    const settle = async (fn: () => void) => {
      if (done) return;
      done = true;
      await cleanup();
      fn();
    };

    void (async () => {
      // Re-probe on every page load and URL change: login is one or more
      // redirects, and we do not want to depend on which page it lands on.
      const runProbe = () => {
        if (!done) void InAppBrowser.executeScript({ code: probe });
      };

      listeners.push(
        await InAppBrowser.addListener("browserPageLoaded", runProbe),
      );
      listeners.push(
        await InAppBrowser.addListener("urlChangeEvent", runProbe),
      );
      listeners.push(
        await InAppBrowser.addListener("closeEvent", () => {
          void settle(() =>
            reject(new ImportCancelled("Student portal sign-in was cancelled.")),
          );
        }),
      );
      listeners.push(
        await InAppBrowser.addListener("messageFromWebview", (event) => {
          const d = event?.detail as
            | { tag?: string; attendanceHtml?: string; marksHtml?: string | null }
            | undefined;
          if (!d || d.tag !== MSG || !d.attendanceHtml) return;
          void settle(async () => {
            try {
              resolve(
                await parseStudentPortal(d.attendanceHtml!, d.marksHtml ?? undefined),
              );
            } catch (e) {
              reject(e as Error);
            }
          });
        }),
      );

      await InAppBrowser.openWebView({
        url: LOGIN_URL,
        title: "Sign in to SRM Student Portal",
        // The portal handles its own captcha and JS; we only read the result.
        isPresentAfterPageLoad: true,
        // Present as an ordinary mobile browser so the portal serves the
        // captcha (see PORTAL_UA). Without this the Android WebView is flagged
        // and the captcha never loads.
        headers: { "User-Agent": PORTAL_UA },
      });
    })().catch((e) => {
      void settle(() => reject(e as Error));
    });
  });
}

File: backend/core/student_portal.py
"""Student portal (sp.srmist.edu.in) support, WebView model.

WHY THIS FILE NO LONGER LOGS IN
-------------------------------
The student portal guards its login with an anti-bot gate: a POST without the
browser fingerprint its JavaScript generates is refused outright with
"JavaScript is required to log into this application securely." Getting a
SCRIPTED login past that means forging that fingerprint (a fake canvas hash,
invented mouse/keystroke counts, a spoofed `navigator.webdriver`). Skipp does
not do that: it is bot-detection evasion, it is brittle (it broke the day it was
tried, 2026-08-17), and it risks the student's own account.

The durable, honest path is a REAL login. Skipp is wrapped with Capacitor, and
the student signs in to the real portal inside an in-app WebView: real page,
real JavaScript, a real human solving the real captcha. Because that WebView is
same-origin with sp.srmist.edu.in once signed in, JavaScript injected into it can
fetch the report pages directly, and the session cookies (including the HttpOnly
JSESSIONID) ride along on their own. The WebView hands the resulting HTML back to
the app, which posts it here to be parsed.

So the backend never sees a password for this source and never signs in. It
receives HTML that a real authenticated session produced and turns it into the
same JSON shape academia gives. The parsers live in `services/sp_attendance.py`
and `services/sp_marks.py`, both written and verified against a real capture.

The report paths are exported for the WebView to fetch. They are POSTed with an
empty body: the servlet reads the student off the session, so there is nothing
to pass and nothing that could be pointed at another student's record.
"""
from __future__ import annotations

BASE_URL = "https://sp.srmist.edu.in"
CONTEXT = "/srmiststudentportal"

#: Where a signed-in student portal WebView fetches each report from.
REPORT_ATTENDANCE = f"{CONTEXT}/students/report/studentAttendanceDetails.jsp"
REPORT_MARKS = f"{CONTEXT}/students/report/studentInternalMarkDetails.jsp"
REPORT_PROFILE = f"{CONTEXT}/students/report/studentProfile.jsp"

#: Markers of a signed-out response. A signed-out report fetch does NOT return
#: the login form: it returns a tiny "login screen is loading" LOADER page that
#: self-submits to youLogin.jsp. So the loader markers matter as much as the
#: login-form ones, or a signed-out fetch reads as "no attendance table"
#: (nothing to parse) rather than the truthful "sign in again".
LOGIN_MARKERS = (
    "SCaptchaServlet",
    'name="password"',
    "LoginServlet",
    "youLogin",
    "loginManager",
    "login screen is loading",
)


class StudentPortalError(Exception):
    """Base class for student portal failures, surfaced to the API as typed 4xx."""


class SessionExpired(StudentPortalError):
    """The HTML we were handed is the login page, not a signed-in report."""


def looks_signed_out(html: str) -> bool:
    """True when this HTML is the login screen rather than a report."""
    return any(marker in html for marker in LOGIN_MARKERS)
